posthog/api/insight.py

```diff
@@ -282,9 +282,45 @@ def calculate_session(self, request: request.Request) -> Dict[str, Any]:
     # ******************************************
     @action(methods=["GET", "POST"], detail=False)
     def funnel(self, request: request.Request, *args: Any, **kwargs: Any) -> Response:
-        result = self.calculate_funnel(request)
-
-        return Response(result)
+        funnel = self.calculate_funnel(request)
+
+        funnel["result"] = self.protect_old_clients_from_multi_property_default(request.data, funnel["result"])
+
+        return Response(funnel)
+
+    @staticmethod
+    def protect_old_clients_from_multi_property_default(
+        data: Dict[str, Any], result: List[List[Dict[str, Any]]]
+    ) -> List[List[Dict[str, Any]]]:
+        """
+        Implementing multi property breakdown will default breakdown to a list even if it is received as a string.
+        This is a breaking change for clients.
+        Clients which do not have multi property breakdown enabled will send a single breakdown as a string
+        This method checks if the request has come in that format and "unboxes" the list
+        to avoid breaking that client
+        :param data: the data in the request
+        :param result: the query result which may contain an unwanted array breakdown
+        :return:
+        """
+        is_single_property_breakdown = (
+            "insight" in data
+            and data["insight"] == "FUNNELS"
+            and "breakdown_type" in data
+            and data["breakdown_type"] in ["person", "event"]
+            and "breakdown" in data
+            and isinstance(data["breakdown"], str)
+        )
+        if is_single_property_breakdown:
+            copied_result = copy.deepcopy(result)
+            for i_index, i in enumerate(result):
+                for j_index, j in enumerate(i):
```

Comment:
I suggest using names more descriptive than `i` and `j`

Reply:
Ha, one place I worked everyone called nested loops "i, j loops" and would have been confused without them. I'll expand them :)